File: 2017/03_SpiralMemory/test/test_spiral.py
# ======================================================================
# Spiral Memory
#   Advent of Code 2017 Day 03 -- Eric Wastl -- https://adventofcode.com
#
# Computer simulation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                      t e s t _ s p i r a l . p y
# ======================================================================
"Test solver for Advent of Code 2017 day 3, Spiral Memory"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import spiral

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

P1_EXAMPLES = [
    (1, 0),
    (6, 1),
    (7, 2),
    (12, 3),
    (23, 2),
    (1024, 31)]

# ======================================================================
#                                                             TestSpiral
# ======================================================================


class TestSpiral(unittest.TestCase):  # pylint: disable=R0904
    """Test Spiral object"""

    def test_manhattan_distance(self):
        """Test default manhattan_distance utility function"""

        # 1. Near the origin
        self.assertEqual(spiral.manhattan_distance((0, 0), (0, 0)), 0)
        self.assertEqual(spiral.manhattan_distance((0, 0), (0, 1)), 1)
        self.assertEqual(spiral.manhattan_distance((0, 0), (-1, 1)), 2)

        # 2. Elsewhere
        self.assertEqual(spiral.manhattan_distance((3, 2), (4, 3)), 2)
        self.assertEqual(spiral.manhattan_distance((3, 0), (3, 1)), 1)
        self.assertEqual(spiral.manhattan_distance((3, 0), (-1, 1)), 5)

    def test_empty_init(self):
        """Test default Spiral creation"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)
        self.assertEqual(myspiral.spiral, {})

        # 3. Check methods
        self.assertEqual(myspiral.steps(1), 0)

    def test_ring_number(self):
        """Test ring_number determination"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)
        self.assertEqual(myspiral.spiral, {})

        # 3. Check ring_number()
        self.assertEqual(myspiral.ring_number(1), 0)
        self.assertEqual(myspiral.ring_number(7), 1)
        self.assertEqual(myspiral.ring_number(10), 2)
        self.assertEqual(myspiral.ring_number(15), 2)
        self.assertEqual(myspiral.ring_number(20), 2)
        self.assertEqual(myspiral.ring_number(25), 2)
        self.assertEqual(myspiral.ring_number(30), 3)
        self.assertEqual(myspiral.ring_number(35), 3)
        self.assertEqual(myspiral.ring_number(40), 3)
        self.assertEqual(myspiral.ring_number(45), 3)
        self.assertEqual(myspiral.ring_number(50), 4)

    def test_ring_side(self):
        """Test length of ring side"""

        # 1. Check ring_side()
        self.assertEqual(spiral.ring_side(0), 1)
        self.assertEqual(spiral.ring_side(1), 3)
        self.assertEqual(spiral.ring_side(2), 5)
        self.assertEqual(spiral.ring_side(3), 7)
        self.assertEqual(spiral.ring_side(4), 9)
        self.assertEqual(spiral.ring_side(21), 43)
        self.assertEqual(spiral.ring_side(30), 61)

    def test_ring_length(self):
        """Test length of ring length"""

        # 1. Check ring_side()
        self.assertEqual(spiral.ring_length(0), 1)
        self.assertEqual(spiral.ring_length(1), 8)
        self.assertEqual(spiral.ring_length(2), 16)
        self.assertEqual(spiral.ring_length(3), 24)
        self.assertEqual(spiral.ring_length(4), 32)
        self.assertEqual(spiral.ring_length(10), 80)
        self.assertEqual(spiral.ring_length(30), 240)

    def test_ring_max_min(self):
        """Test highest and lowest number on a ring"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)
        self.assertEqual(myspiral.spiral, {})

        # 3. Check ring_max()
        self.assertEqual(myspiral.ring_max(0), 1)
        self.assertEqual(myspiral.ring_max(1), 9)
        self.assertEqual(myspiral.ring_max(2), 25)
        self.assertEqual(myspiral.ring_max(3), 49)
        self.assertEqual(myspiral.ring_max(4), 81)
        self.assertEqual(myspiral.ring_max(5), 121)

        # 3. Check ring_min()
        self.assertEqual(myspiral.ring_min(0), 1)
        self.assertEqual(myspiral.ring_min(1), 2)
        self.assertEqual(myspiral.ring_min(2), 10)
        self.assertEqual(myspiral.ring_min(3), 26)
        self.assertEqual(myspiral.ring_min(4), 50)
        self.assertEqual(myspiral.ring_min(5), 82)

    def test_ring_offsets(self):
        """Test ring and side offsets"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)
        self.assertEqual(myspiral.spiral, {})

        # 3. Check ring_offset()
        self.assertEqual(myspiral.ring_offset(0), None)
        self.assertEqual(myspiral.ring_offset(1), 0)
        self.assertEqual(myspiral.ring_offset(5), 4)
        self.assertEqual(myspiral.ring_offset(10), 1)
        self.assertEqual(myspiral.ring_offset(15), 6)
        self.assertEqual(myspiral.ring_offset(20), 11)
        self.assertEqual(myspiral.ring_offset(25), 0)
        self.assertEqual(myspiral.ring_offset(30), 5)

        # 3. Check side_offset()
        self.assertEqual(myspiral.side_offset(0), None)
        self.assertEqual(myspiral.side_offset(1), 0)
        self.assertEqual(myspiral.side_offset(5), 0)
        self.assertEqual(myspiral.side_offset(10), 1)
        self.assertEqual(myspiral.side_offset(15), 2)
        self.assertEqual(myspiral.side_offset(20), 3)
        self.assertEqual(myspiral.side_offset(25), 0)
        self.assertEqual(myspiral.side_offset(30), 5)

    def test_side_distance(self):
        """Test distance to horizontal or vertical axis"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)
        self.assertEqual(myspiral.spiral, {})

        # 3. Check ring_offset()
        self.assertEqual(myspiral.side_distance(0), None)
        self.assertEqual(myspiral.side_distance(1), 0)
        self.assertEqual(myspiral.side_distance(5), 1)
        self.assertEqual(myspiral.side_distance(10), 1)
        self.assertEqual(myspiral.side_distance(15), 0)
        self.assertEqual(myspiral.side_distance(20), 1)
        self.assertEqual(myspiral.side_distance(25), 2)
        self.assertEqual(myspiral.side_distance(30), 2)

    def test_part_one_examples(self):
        """Test Spiral examples"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)

        # 3. Loop for all of the part 1 examples
        for example in P1_EXAMPLES:

            # 4. Check the number of steps from the square to the access port
            self.assertEqual(myspiral.steps(example[0],
                                            verbose=False),
                             example[1])

    def test_part_one_spiral(self):
        """Test building the part 1 spiral"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral()

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, False)

        # 3. Build the spiral
        result = myspiral.build_until_gt(23, verbose=False)

        # 4. Validate the spiral
        self.assertEqual(result, 24)
        self.assertEqual(myspiral.spiral[(0, 0)], 1)
        self.assertEqual(myspiral.spiral[(1, 0)], 2)
        self.assertEqual(myspiral.spiral[(1, 1)], 3)
        self.assertEqual(myspiral.spiral[(0, 1)], 4)
        self.assertEqual(myspiral.spiral[(-1, 1)], 5)
        self.assertEqual(myspiral.spiral[(-1, 0)], 6)
        self.assertEqual(myspiral.spiral[(-1, -1)], 7)
        self.assertEqual(myspiral.spiral[(-1, -2)], 22)
        self.assertEqual(myspiral.spiral[(0, -2)], 23)
        self.assertEqual(myspiral.spiral[(1, -2)], 24)
        self.assertTrue((2, -2) not in myspiral.spiral)

    def test_part_two_spiral(self):
        """Test building the part 2 spiral"""

        # 1. Create default Spiral object
        myspiral = spiral.Spiral(part2=True)

        # 2. Make sure it has the default values
        self.assertEqual(myspiral.part2, True)

        # 3. Build the spiral
        result = myspiral.build_until_gt(806, verbose=False)

        # 4. Validate the spiral
        self.assertEqual(result, 880)
        self.assertEqual(myspiral.spiral[(0, 0)], 1)
        self.assertEqual(myspiral.spiral[(1, 0)], 1)
        self.assertEqual(myspiral.spiral[(1, 1)], 2)
        self.assertEqual(myspiral.spiral[(0, 1)], 4)
        self.assertEqual(myspiral.spiral[(-1, 1)], 5)
        self.assertEqual(myspiral.spiral[(-1, 0)], 10)
        self.assertEqual(myspiral.spiral[(-1, -1)], 11)
        self.assertEqual(myspiral.spiral[(-1, -2)], 747)
        self.assertEqual(myspiral.spiral[(0, -2)], 806)
        self.assertEqual(myspiral.spiral[(1, -2)], 880)
        self.assertTrue((2, -2) not in myspiral.spiral)

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                     t e s t _ s p i r a l . p y                end
# ======================================================================

File: 2017/03_SpiralMemory/test_spiral.lua
-- ======================================================================
-- SpiralMemory
--   Advent of Code 2017 Day 03 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ s p i r a l . l u a
-- ======================================================================
-- Test solver for Advent of Code 2017 day 03, SpiralMemory

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local spiral = require('spiral')

-- ----------------------------------------------------------------------
--                                                              from_text
-- ----------------------------------------------------------------------

function from_text(text)
  -- Break the text into trimed, non-comment lines

  -- 1. We start with no lines
  local result = {}

  -- 2. Loop for lines in the text
  for line in text:gmatch('[^\r\n]+') do

    -- 3. But ignore blank and comment lines
    line = line:gsub("%s*$", "")
    if #line > 0 and "!" ~= line:sub(1, 1) then

      -- 4. Add the line
      table.insert(result, line)
    end
  end

  -- 5. Return a table of cleaned lines
  return result
end

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = [[
1024
]]
local PART_ONE_TEXT = EXAMPLE_TEXT
local PART_TWO_TEXT = EXAMPLE_TEXT

local PART_ONE_RESULT = 31
local PART_TWO_RESULT = 1968

-- ======================================================================
--                                                             TestSpiral
-- ======================================================================

function test_empty_init()
  -- Test the default Spiral creation

  -- 1. Create default Spiral object
  local myobj = spiral:Spiral()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(myobj.value, 0)
  luaunit.assertEquals(myobj.x, 0)
  luaunit.assertEquals(myobj.y, 0)
  luaunit.assertEquals(myobj.dir, "E")
  luaunit.assertEquals(myobj.steps, 1)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:location(), "0,0")
  luaunit.assertEquals(myobj:distance(), 0)
  myobj:grow()
  luaunit.assertEquals(myobj.x, 1)
  luaunit.assertEquals(myobj.y, 0)
  luaunit.assertEquals(myobj.dir, "N")
  luaunit.assertEquals(myobj.steps, 2)
  luaunit.assertEquals(myobj:location(), "1,0")
  luaunit.assertEquals(myobj:distance(), 1)
  myobj:grow()
  luaunit.assertEquals(myobj.x, 1)
  luaunit.assertEquals(myobj.y, 1)
  luaunit.assertEquals(myobj.dir, "W")
  luaunit.assertEquals(myobj.steps, 3)
  luaunit.assertEquals(myobj:location(), "1,1")
  luaunit.assertEquals(myobj:distance(), 2)
  myobj:grow()
  luaunit.assertEquals(myobj.x, 0)
  luaunit.assertEquals(myobj.y, 1)
  luaunit.assertEquals(myobj.dir, "W")
  luaunit.assertEquals(myobj.steps, 4)
  luaunit.assertEquals(myobj:location(), "0,1")
  luaunit.assertEquals(myobj:distance(), 1)
  myobj:grow()
  luaunit.assertEquals(myobj.x, -1)
  luaunit.assertEquals(myobj.y, 1)
  luaunit.assertEquals(myobj.dir, "S")
  luaunit.assertEquals(myobj.steps, 5)
  luaunit.assertEquals(myobj:location(), "-1,1")
  luaunit.assertEquals(myobj:distance(), 2)
  

end

function test_text_init()
  -- Test the Spiral object creation from text

  -- 1. Create Spiral object from text
  local myobj = spiral:Spiral({text=from_text(EXAMPLE_TEXT)})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 1)
  luaunit.assertEquals(myobj.value, 1024)
  luaunit.assertEquals(myobj.x, 0)
  luaunit.assertEquals(myobj.y, 0)
  luaunit.assertEquals(myobj.dir, "E")
  luaunit.assertEquals(myobj.steps, 1)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:grow_until(12), 3)
  luaunit.assertEquals(myobj:grow_until(23), 2)
  luaunit.assertEquals(myobj:grow_until(1024), 31)

end

function test_two_init()
  -- Test the Spiral object creation from text

  -- 1. Create Spiral object from text
  local myobj = spiral:Spiral({text=from_text(EXAMPLE_TEXT), part2=true})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, true)
  luaunit.assertEquals(#myobj.text, 1)
  luaunit.assertEquals(myobj.value, 1024)
  luaunit.assertEquals(myobj.x, 0)
  luaunit.assertEquals(myobj.y, 0)
  luaunit.assertEquals(myobj.dir, "E")
  luaunit.assertEquals(myobj.steps, 1)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:location(), "0,0")
  luaunit.assertEquals(myobj.grid["0,0"], 1)
  myobj:grow()
  luaunit.assertEquals(myobj:location(), "1,0")
  luaunit.assertEquals(myobj.grid["1,0"], 1)
  myobj:grow()
  luaunit.assertEquals(myobj:location(), "1,1")
  luaunit.assertEquals(myobj.grid["1,1"], 2)

  luaunit.assertEquals(myobj:grow_until(12), 23)
  luaunit.assertEquals(myobj:grow_until(23), 25)
  luaunit.assertEquals(myobj:grow_until(1024), 1968)

end
function test_part_one()
  -- Test part one example of Spiral object

  -- 1. Create Spiral object from text
  local myobj = spiral:Spiral({text=from_text(PART_ONE_TEXT)})

  -- 2. Check the part one result
  luaunit.assertEquals(myobj:part_one({verbose=false, limit=0}), PART_ONE_RESULT)

end

function test_part_two()
  -- Test part two example of Spiral object

  -- 1. Create Spiral object from text
  local myobj = spiral:Spiral({part2=true, text=from_text(PART_TWO_TEXT)})

  -- 2. Check the part two result
  luaunit.assertEquals(myobj:part_two({verbose=false, limit=0}), PART_TWO_RESULT)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                  t e s t _ s p i r a l . l u a                 end
-- ======================================================================
